pkg/api/validate/dynamic/dynamic.go

```diff
@@ -197,27 +197,46 @@ func (dv *dynamic) validateVnetPermissions(ctx context.Context, vnet azure.Resou
 		"Microsoft.Network/virtualNetworks/subnets/write",
 	})
 
+	noPermissionsErr := api.NewCloudError(
+		http.StatusBadRequest,
+		errCode,
+		"",
+		"The %s service principal does not have Network Contributor permission on vnet '%s'.",
```

Comment:
```suggestion
		"The %s service principal does not have Network Contributor role on vnet '%s'.",
```

Reply:
is it possible to include the application ID? say, if `authorizerType` is `AuthorizerFirstParty`, the user still wouldn't know what app ID to give role to, right?

Reply:
done.  changed from permission to role and included the appID field which is printed in the error message.